### ISSUE TITLE
srt_epoll_uwait results in SRT_EPOLL_OUT even after a client is disconnected  

### ISSUE THREAD

**Issue by ShinyMacadamia** [Haivision/srt#3061]
Consider this example: 
```cpp
#include <netdb.h>
#include <srt.h>

#include <cstdlib>
#include <cstring>
#include <fstream>
#include <iostream>
#include <string>

using namespace std;

int main(int argc, char* argv[]) {
    if ((2 < argc) || ((2 == argc) && (0 == atoi(argv[1])))) {
        cout << "usage: recvlive [server_port]" << endl;
        return 0;
    }

    srt_startup();
    srt_setloglevel(srt_logging::LogLevel::debug);

    addrinfo hints;
    addrinfo* res;

    memset(&hints, 0, sizeof(struct addrinfo));
    hints.ai_flags = AI_PASSIVE;
    hints.ai_family = AF_INET;
    hints.ai_socktype = SOCK_DGRAM;

    string service("9000");
    if (2 == argc) service = argv[1];

    if (0 != getaddrinfo(NULL, service.c_str(), &hints, &res)) {
        cout << "illegal port number or port is busy.\n" << endl;
        return 0;
    }

    SRTSOCKET sfd = srt_create_socket();
    if (SRT_INVALID_SOCK == sfd) {
        cout << "srt_socket: " << srt_getlasterror_str() << endl;
        return 0;
    }

    bool no = false;
    if (SRT_ERROR == srt_setsockopt(sfd, 0, SRTO_RCVSYN, &no, sizeof no)) {
        cout << "srt_setsockopt: " << srt_getlasterror_str() << endl;
        return 0;
    }

    SRT_TRANSTYPE trans_type = SRTT_LIVE;
    if (SRT_ERROR == srt_setsockopt(sfd, 0, SRTO_TRANSTYPE, &trans_type,
                                    sizeof trans_type)) {
        cout << "srt_setsockopt: " << srt_getlasterror_str() << endl;
        return 0;
    }

    if (SRT_ERROR == srt_bind(sfd, res->ai_addr, res->ai_addrlen)) {
        cout << "srt_bind: " << srt_getlasterror_str() << endl;
        return 0;
    }

    // Start listening
    if (SRT_ERROR == srt_listen(sfd, 10)) {
        cout << "srt_listen: " << srt_getlasterror_str() << endl;
        return 0;
    }

    int epoll_flags = SRT_EPOLL_IN | SRT_EPOLL_OUT | SRT_EPOLL_ERR;

    int epid = srt_epoll_create();
    srt_epoll_add_usock(epid, sfd, &epoll_flags);

    SRT_EPOLL_EVENT epoll_results[2];

    for (;;) {
        int epoll_res = srt_epoll_uwait(epid, epoll_results, 3, 500);

        for (int i = 0; i < epoll_res; i++) {
            SRTSOCKET sock = epoll_results[i].fd;
            auto event = epoll_results[i].events;

            if (sock == sfd &&
                event == SRT_EPOLL_OUT) {  // triggers even after the client got
                                           // disconnect.
                continue;
            }

            if (event & SRT_EPOLL_ERR) {
                std::cout << "epoll error detected, removing socket"
                          << std::endl;

                srt_epoll_remove_usock(epid, sock);
                srt_close(sock);
                continue;
            }

            if (event & SRT_EPOLL_IN) {
                if (sock == sfd) {
                    SRTSOCKET client_sock;
                    sockaddr_storage client_addr;
                    int addrlen = sizeof(client_addr);

                    client_sock =
                        srt_accept(sfd, (sockaddr*)&client_addr, &addrlen);
                    if (client_sock == SRT_INVALID_SOCK) {
                        cout << "srt_accept: " << srt_getlasterror_str()
                             << endl;
                        continue;
                    }

                    int events = SRT_EPOLL_ERR;
                    srt_epoll_add_usock(epid, client_sock, &events);

                    cout << "New client connected" << endl;
                }
            }
        }
    }
}
```
Using this snipped (based on the API documentation and examples) i get this weird (albeit expected?) behavior:
1) When the application only starts and there are no clients connected, the `srt_epoll_uwait` function always returns `0`. 
2) When a client connects (just 1 client for simplicity sake) it returns 1 and results in `SRT_EPOLL_IN` event status (which is expected). 
3) After that `srt_epoll_uwait` results in 1 and `SRT_EPOLL_OUT` event status. 
4) But when the client disconnects it once again results in `SRT_EPOLL_OUT`

Is it expected that waiting on a listener socket always results in `SRT_EPOLL_OUT` after the very first connection accepted ? Even if a client got disconnected? The problem with such behavior is that it creates an infinite loop which results in high cpu usage.

As far as i currently understand the only way to prohibit such behavior is to remove `SRT_EPOLL_OUT` event via `srt_epoll_update_` when the number of clients becomes 0. Is there a better way? 

Thanks in advance!

**Comment by ShinyMacadamia** [comment]
> Expected is getting `SRT_EPOLL_ERR` on disconnect.

> All of IN, OUT and ERR are set in this case. This is because if you don't poll on ERR, at least some signal must be given to the application that it should perform the operation, even though the system knows that the operation will fail, and this is expected.

Sorry, folks, i missed this part. Yes, polling indeed returns an error in case a client disconnects. The problem is that the listener socket always yields OUT even after all clients are disconnected. On top of that i noticed that polling OUT events is expensive due to frequency at which such polling happens (basically an infinite loop). My single core cpu usage goes up to 100% (i7 14700k) in case i poll for OUT events -  ` int epoll_flags = SRT_EPOLL_IN | SRT_EPOLL_OUT | SRT_EPOLL_ERR;`

**Comment by maxsharabayko** [comment]
Try edge-triggered events, see [SRT_EPOLL_ET](https://github.com/Haivision/srt/blob/master/docs/API/API-functions.md#srt_epoll_update_ssock).

**Comment by ethouris** [comment]
OUT event on a listener socket should never happen. We found that this happens in some older version, this is an old bug that was overlooked for a long time (it's present even in UDT). I don't remember when exactly it was fixed, but it shouldn't be present anymore in the latest version.

The listener socket sets the IN flag in case when there's a new connection pending for being accepted. I don't even remember if there are any circumstances where it sets the ERR flag. OUT flag should not be set and even if it happened in some earlier version that this is set, it's a mistake and the application should not subscribe to this flag on a listener socket.

As per normal, "transmission" socket, note that the OUT flag is normally always set on a socket, unless you have overflown the sender buffer with data by scheduling them faster than the network can send them. This is the only case when OUT flag can be cleared, otherwise it is set all the time. If you are sending a live stream, it is usually already dampened enough so that this shall never happen, so if you want on an EID that is subscribed on OUT flag on a socket, it will for 99% of cases exit immediately. So it's usually a bad idea to have IN and OUT flag subscribed in the same EID. The latter should be at best subscribed on demand when you have experienced sending failure with SRT_EASYNCSND and you want to wait until sending is possible again. And of course on a different EID than the one that you use for waiting for data. If you use a single thread and therefore you might have a situation that there's nothing to read (no IN flag) and no place in the sender buffer (no OUT flag) and you want to wait for one of these things to happen, this should be done only in an ususual situation when you do have data to send, you can't send them now (because OUT flag is clear) and you don't know if you have data to read. For that special case you can do something like subscribing for OUT when you detect that choking situation so that you wait for any of IN and OUT, but then unsubscribe once you detected that OUT flag is set again.

I haven't analyzed your code so I don't know if you are using bidirectional transmission, but that's the only case when you might want to do tricks I described above. If you use a single direction for sending, simply subscribe only to that direction. Listener sockets is another matter - you should subscribe only for IN to get information about the incoming connection.

**Comment by ShinyMacadamia** [comment]
> OUT event on a listener socket should never happen. We found that this happens in some older version, this is an old bug that was overlooked for a long time (it's present even in UDT). I don't remember when exactly it was fixed, but it shouldn't be present anymore in the latest version.
> 
> The listener socket sets the IN flag in case when there's a new connection pending for being accepted. I don't even remember if there are any circumstances where it sets the ERR flag. OUT flag should not be set and even if it happened in some earlier version that this is set, it's a mistake and the application should not subscribe to this flag on a listener socket.
> 
> As per normal, "transmission" socket, note that the OUT flag is normally always set on a socket, unless you have overflown the sender buffer with data by scheduling them faster than the network can send them. This is the only case when OUT flag can be cleared, otherwise it is set all the time. If you are sending a live stream, it is usually already dampened enough so that this shall never happen, so if you want on an EID that is subscribed on OUT flag on a socket, it will for 99% of cases exit immediately. So it's usually a bad idea to have IN and OUT flag subscribed in the same EID. The latter should be at best subscribed on demand when you have experienced sending failure with SRT_EASYNCSND and you want to wait until sending is possible again. And of course on a different EID than the one that you use for waiting for data. If you use a single thread and therefore you might have a situation that there's nothing to read (no IN flag) and no place in the sender buffer (no OUT flag) and you want to wait for one of these things to happen, this should be done only in an ususual situation when you do have data to send, you can't send them now (because OUT flag is clear) and you don't know if you have data to read. For that special case you can do something like subscribing for OUT when you detect that choking situation so that you wait for any of IN and OUT, but then unsubscribe once you detected that OUT flag is set again.
> 
> I haven't analyzed your code so I don't know if you are using bidirectional transmission, but that's the only case when you might want to do tricks I described above. If you use a single direction for sending, simply subscribe only to that direction. Listener sockets is another matter - you should subscribe only for IN to get information about the incoming connection.

Thanks for the tips!!! 
About the OUT event on the listener. I'm compiling right from the master branch. And the OUT events on the listener are still present (not sure whether there might be a problem with my own snippet or not).

**Comment by ethouris** [comment]
Sorry, you are right. This fix #2732 was postponed to 1.6.0. Should be available soon in the latest master as long as the new release is kicked off.

**Comment by ShinyMacadamia** [comment]
Thank you everyone participated! Can i close this issue ?

**Comment by maxsharabayko** [comment]
@ShinyMacadamia sure

**Comment by maxsharabayko** [comment]
Closed. Fix #2732 is planned for 1.6.0.
